### PR TITLE
Improved adding routes

### PR DESCRIPTION
**Pull request by mchf**
[bsc#927629](https://bugzilla.suse.com/show_bug.cgi?id=927629)  "yast lan does not allow IPv6 and CIDR subnet masks in static routes"

### CODE REVIEW

File: src/include/network/services/routing.rb [hunk 1]
```diff
@@ -140,6 +140,23 @@ def initialize_network_services_routing(include_target)
...
+    # prefix length. If prefix length is used it has to start with '/'.
+    # For IPv6 network, only prefix length is allowed.
+    #
+    # param [string] netmask or /<prefix length>
+    def valid_netmask?(netmask)
+      return false if netmask.nil? || netmask.empty?
+      return true if Netmask.Check4(netmask)
+
+      prefix = netmask[1..-1].to_i
+      return true if netmask.start_with?("/") && prefix.between?(1, 128)
```
Comment (jreidinger) [inline src/include/network/services/routing.rb]: well, I think this code is more readable, as prefix is needed only if it start with "/"
```
if netmask.start_with?("/")
  return netmask[1..-1].to_i.between?(1, 128)
end

false
```

File: src/include/network/lan/help.rb [hunk 2]
```diff
@@ -98,8 +98,12 @@ def initialize_network_lan_help(_include_target)
                                ) +
                                  _(
                                    "<p>For each route, enter destination network IP address, gateway address,\n" \
-                                     "and netmask. To omit any of these values, use a dash sign \"-\". Select\n" \
-                                     "the device through which the traffic to the defined network will be routed.\"-\" is an alias for any interface.</p>\n"
+                                     "and netmask. You can use either IPv4 netmask or prefix lenght when defining\n" \
```
Comment (mchf) [inline src/include/network/lan/help.rb]: If that is the only grammar issue, I have to congratulate myself ;-) 

Thanks, fixing.

File: src/include/network/services/routing.rb [hunk 3]
```diff
@@ -258,7 +271,7 @@ def RoutingEditDialog(id, entry, devs)
         end
         route = Builtins.add(route, val)
         val = Convert.to_string(UI.QueryWidget(Id(:genmask), :Value))
-        if val != "-" && val != "0.0.0.0" && !Netmask.Check(val)
+        if val != "-" && val != "0.0.0.0" && !valid_netmask?(val)
```
Comment (mvidner) [inline src/include/network/services/routing.rb]: How am I supposed to know the difference between "check the netmask" and "is the netmask valid"?
Why are we not using/fixing the Netmask library instead? Observe these:
- http://www.rubydoc.info/github/yast/yast-yast2/Yast/NetmaskClass
- http://www.rubydoc.info/github/yast/yast-network/Yast/NetworkLanAddressInclude:ValidateNetmask
- http://www.rubydoc.info/github/yast/yast-network/Yast/NetworkRoutinesInclude:validPrefixOrNetmask

File: src/include/network/services/routing.rb [hunk 4]
```diff
@@ -424,43 +437,65 @@ def validateRouting(_key, _event)
       false
     end
 
+    # Mangles route definition for storing in routes file
```
Comment (mvidner) [inline src/include/network/services/routing.rb]: "Mangling" is not something you want to introduce. See http://www.thefreedictionary.com/mangle (C++ mangling is, well, C++, not a good role model)
"Adapt" or "convert".

File: src/include/network/services/routing.rb [hunk 5]
```diff
@@ -424,43 +438,65 @@ def validateRouting(_key, _event)
...
+      netmask = route["netmask"]
+
+      if Netmask.Check4(netmask)
+        cidr = Netmask.ToBits(netmask)
+      elsif netmask.start_with?("/")
+        cidr = netmask[1..-1]
+      else
+        # if it is netmask then long netmask is not supported for IPv6
+        # if it is prefix length (CIDR), then prefix it has to be prefixed by '/'
+        raise ArgumentError, "Invalid netmask or prefix length"
```
Comment (jreidinger) [inline src/include/network/services/routing.rb]: it is nice to add value to exception, so immediatelly visible like `"Invalid netmask or prefix length: '#{netmask}'"

File: src/include/network/services/routing.rb [hunk 6]
```diff
@@ -424,43 +438,65 @@ def validateRouting(_key, _event)
...
+        "netmask"     => "-",
+        "device"      => @defgwdev
+      } if !@defgw.empty?
+
+      route_conf << {
+        "destination" => "default",
+        "gateway"     => @defgw6,
+        "netmask"     => "-",
+        "device"      => @defgwdev6
+      } if !@defgw6.empty?
```
Comment (jreidinger) [inline src/include/network/services/routing.rb]: just question, why @defgw6 and others are instance variables and not local ones?

File: src/modules/Routing.rb [hunk 7]
```diff
@@ -205,6 +205,35 @@ def WriteIPForwarding
...
+
+        if !prefix.nil?
+          route["destination"] = subnet
+          route["netmask"] = "/#{prefix}"
+        end
+
+        route
+      end
+
+      routes
```
Comment (jreidinger) [inline src/modules/Routing.rb]: Very nasty sideeffect as it looks like it do nothing as you do not assign map result to routes variable or use `map!`. But in fact change hash in map, so even each do that stuff. Also it is not clear that it modify parameter itself.

File: src/include/network/services/routing.rb (hunk 6 above)
Comment (mchf) [inline src/include/network/services/routing.rb]: most probably bcs of conversion from ycp.

File: src/include/network/services/routing.rb (hunk 5 above)
Comment (mchf) [inline src/include/network/services/routing.rb]: ok

File: src/include/network/services/routing.rb (hunk 1 above)
Comment (mchf) [inline src/include/network/services/routing.rb]: ok

**Comment by jreidinger** [comment]
LGTM

File: src/modules/Routing.rb [hunk 8]
```diff
@@ -205,6 +205,35 @@ def WriteIPForwarding
       write_ipv6_forwarding(@Forward_v6)
     end
 
+    # Converts routes config as read from system into well-defined format
```
Comment (mvidner) [inline src/modules/Routing.rb]: UGH! Why do we introduce both `normalize_routes(routes)` and `convert_route_conf(route)`?

**Comment by mchf** [comment]
overall comment, for reference;
Purpose of this PR is to 
1) allow creating route using CIDR notation
2) normalize routes representation. Basically "netmask" field is obsolete in routes conf and CIDR should be preferred.
3) follow convention when using netmask in IPv6 networks is inconvenient and prefix should be used there.

(1) is the reported bug, (2) and (3) is covered e.g. by "man routes"

File: src/include/network/services/routing.rb (hunk 3 above)
Comment (mchf) [inline src/include/network/services/routing.rb]: bcs the input value need not be only netmask but also prefix. I agree that ```valid_netmask?``` is most probably not the best name here. I've used "netmask" as a reference to the column name which is being validated.

File: src/include/network/services/routing.rb (hunk 4 above)
Comment (mchf) [inline src/include/network/services/routing.rb]: ok

File: src/modules/Routing.rb (hunk 8 above)
Comment (mchf) [inline src/modules/Routing.rb]: apart of that ```normalize_routes``` works on top of an array of routes, the basic difference is that ```normalize_routes``` should turn route definition into form which is expected in ```convert_route_conf``` - that "destination" field carries only network ip and netmask or prefix is placed in "destination"

**Comment by mvidner** [comment]
We are adding two methods to change the routing settings, to convert between two formats of these, but we have no documentation of even one of these formats.

Forget the legacy code for a while. If we design the model class like follows,
we can say: `convert_route_conf` changes the hash so that raw_netmask is "-"
and `normalize_routes` changes the hash so that raw_netmask contains
a quad or a slash-number.

Or do it some other way, but do have one place that has a Yardoc URL where I can see what a routing entry is. 

```rb
class Yast::New::Routing      # I cannot remember the new namespacing style :(
  # @return [Array<RoutingEntry>]
  attr :entries
  # @return [Boolean]
  attr :forward_ipv4
  # @return [Boolean]
  attr :forward_ipv6

  class RoutingEntry
    # @return [String]  eg. "10.20.0.0"
    def destination_pfx; end
    # @return [Integer] eg. 16
    def destination_len; end

    # @return [String] always in the form "10.20.0.0/16", or "default"
    def destination; end

    # either as length or as a mask, after a slash or as a separate argument
    # destination = "default"
    # destination = "10.20.0.0/16"
    # destination = "10.20.0.0/255.255.0.0"
    # destination = ["10.20.0.0", 16]
    # destination = ["10.20.0.0", "255.255.0.0"]
    def destination=(pfx, len_or_mask=nil); end

    # @return [String]
    attr :gateway
    # @return [String] "eth0"
    attr :device
    # @return [String]
    attr :extrapara

    # @param [Hash{String=>String}] h the keys are
    #   "destination", "gateway", "netmask", "device", "extrapara"
    def self.from_anyagent(h); end
    # @return [Hash{String=>String}] h the keys are
    #   "destination", "gateway", "netmask", "device", "extrapara"
    def to_anyagent; end

    private

    # @return [String] The raw value of the 1st column of routes(5)
    def raw_destination; end
    # @return [String] The raw value of the 3rd column of routes(5)
    def raw_netmask; end
  end
end
```

**Comment by mchf** [comment]
nice example, I'll go through it. However, I'm afraid that when going this way I cannot handle it as a part of "Kill as many bugs as possible in a given time" task ;-)

**Comment by mvidner** [comment]
If we want to get this over with quickly, I suggest this option:

>  do have one place that has a Yardoc URL where I can see what a routing entry is.

**Comment by mvidner** [comment]
@teclator do you have an opinion on this?

**Comment by teclator** [comment]
In case that we accept it for the quick bug fixing I would consider the @mvidner suggestion as soon as possible. We have some bugs related to static routes created/applied by YaST as for example https://bugzilla.suse.com/show_bug.cgi?id=1005478 so could be a good oportunity to improve current code.

**Comment by mvidner** [comment]
@mchf can you pull in https://github.com/yast/yast-network/commit/7667199617851d330714b12d1109cc79ba9e48ae from https://github.com/yast/yast-network/tree/bnc927629-routes-ipv6 please?

**Comment by mchf** [comment]
@mvidner pulled, rebased

**Pull request merged**